File: src/main/java/june_2025/CourseSchedule.java
package june_2025;

import java.util.*;

public class CourseSchedule {

    public static void main(String[] args) {
        int numCourses = 4;
        int[][] prerequisites = {{1,2}, {0, 1}, {2,0}, {3,1}, {3,2}};

        CourseSchedule courseSchedule = new CourseSchedule();
        boolean result = courseSchedule.canFinish(numCourses, prerequisites);
        System.out.println(result);
    }

    // O(V + E) time | O(V + E) space
    public boolean canFinish(int numCourses, int[][] prerequisites) {
        Map<Integer, List<Integer>> adj = new HashMap<>();
        for (int i = 0; i < numCourses; i++) {
            adj.put(i, new ArrayList<>());
        }
        for (int[] pre : prerequisites) {
            int s = pre[0];
            int d = pre[1];
            List<Integer> sNode = adj.get(s);
            sNode.add(d);
        }

        Set<Integer> visited = new HashSet<>();
        Set<Integer> visiting = new HashSet<>();

        for (int c = 0; c < numCourses; c++) {
            if (!visited.contains(c)) {
                if (cycle(adj, c, visiting, visited)) {
                    return false;
                }
            }
        }
        return true;
    }

    private boolean cycle(Map<Integer, List<Integer>> adj, int node, Set<Integer> visiting, Set<Integer> visited) {
        if (visited.contains(node)) {
            return false;
        }
        if (visiting.contains(node)) {
            return true;
        }
        visiting.add(node);

        List<Integer> neighbors = adj.get(node);
        for (int nei : neighbors) {
            if (cycle(adj, nei, visiting, visited)) {
                return true;
            }
        }

        visiting.remove(node);
        visited.add(node);
        return false;
    }

}

File: python/course_schedule.py
class Solution(object):
    # O(V + E) time | O(V + E) space
    def canFinish(self, numCourses, prerequisites):
        """
        :type numCourses: int
        :type prerequisites: List[List[int]]
        :rtype: bool
        """
        adj = {}
        for i in range(numCourses):
            adj[i] = []
        
        for pre in prerequisites:
            s = pre[0]
            d = pre[1]
            sNode = adj[s]
            sNode.append(d)
        
        def cycle(node):
            if node in visited:
                return False
            if node in visiting:
                return True
            
            visiting.add(node)
            neighbors = adj[node]

            for nei in neighbors:
                if cycle(nei):
                    return True

            visiting.remove(node)
            visited.add(node)
            return False


        visited = set()
        visiting = set()
        for c in range(numCourses):
            if not c in visited:
                if cycle(c):
                    return False
        return True
    
# prerequisites = [[1,2], [0, 1], [2,0], [3,1], [3,2]]   
prerequisites = [[1,0], [2,0], [3,1], [3,2]]    
numCourses = 4

solution = Solution()
print(solution.canFinish(numCourses, prerequisites))